### PR TITLE
[release/6.0] Throw for bad requests intended for minimal route handlers in development

### PR DESCRIPTION
**Pull request by halter73**
Backport of #36004 to release/6.0

## Customer Impact

Now instead of having to enable debug logging in development to see why a request is failing with a 400 response, you can easily see why the request is being rejected using the developer exception page or error logs.

This updates the DeveloperExceptionPageMiddleware to preserve the status code requested by the BadHttpRequestException instead of always responding with a 500. This way, the client can still get a 400 response in development but also get more error details.

This also updates the log messages (and the matching Exception messages) to use TypeNameHelper to produce friendlier parameter type names. And this includes the changes from https://github.com/dotnet/aspnetcore/pull/35580 which I'll close in favor of this.

image

## Testing

This adds a lot of new unit and functional tests to cover this scenario. I also tested manually.

## Risk

Low. This only affects bad requests that would be rejected either way. This also only impacts development environments to make debugging easier unless you opt-in.

### CODE REVIEW

**Comment by halter73** [comment]
Sorry for the code review span. I accidently targeted `main` as the base branch initially.

File: src/Http/Routing/src/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -14,6 +14,10 @@ Microsoft.AspNetCore.Routing.IDataTokensMetadata.DataTokens.get -> System.Collec
 Microsoft.AspNetCore.Routing.IRouteNameMetadata.RouteName.get -> string?
 Microsoft.AspNetCore.Routing.Matching.IParameterLiteralNodeMatchingPolicy
 Microsoft.AspNetCore.Routing.Matching.IParameterLiteralNodeMatchingPolicy.MatchesLiteral(string! parameterName, string! literal) -> bool
+Microsoft.AspNetCore.Routing.RouteHandlerOptions
+Microsoft.AspNetCore.Routing.RouteHandlerOptions.RouteHandlerOptions() -> void
+Microsoft.AspNetCore.Routing.RouteHandlerOptions.ThrowOnBadRequest.get -> bool
+Microsoft.AspNetCore.Routing.RouteHandlerOptions.ThrowOnBadRequest.set -> void
```
Comment (dougbu) [inline src/Http/Routing/src/PublicAPI.Unshipped.txt:20]: Was the new `class` API reviewed for 6.0 RC2❔

Comment (halter73) [reply src/Http/Routing/src/PublicAPI.Unshipped.txt:20]: There's an email API review. I just cc'd you.

**Review by davidfowl — APPROVED** [review]

**Comment by halter73** [comment]
@dotnet/aspnet-build Can someone please merge this? Thanks!

**Pull request merged**